[MSE] Improved coded frame eviction algorithm

calvaris opened the pull request:
#### 48b51f0536c90a12e3acd596d36ef82324acd3a5
<pre>
[MSE] Improved coded frame eviction algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=246144">https://bugs.webkit.org/show_bug.cgi?id=246144</a>

Reviewed by Alicia Boya Garcia.

I made it in two aspects, first is that now it does not begin at 0, it
begins at the first buffered range.

A fix would be that sometimes you need to evict frames but a window of
30s is too much and it can easily happen it does not evict anything,
so we need to reduce the partition to be able to effectively evict
something. We do it in windows from 30 to 3s.

This is tested by several tests that needed a rebase line or small tweaks.

* LayoutTests/media/media-source/media-source-append-before-last-range-no-quota-exceeded-expected.txt:
* LayoutTests/media/media-source/media-source-append-before-last-range-no-quota-exceeded.html:
* LayoutTests/media/media-source/media-source-append-buffer-full-quota-exceeded-error-expected.txt:
* LayoutTests/media/media-source/media-source-append-buffer-full-quota-exceeded-error.html:
* LayoutTests/media/media-source/media-source-evict-codedframe-after-seek-expected.txt:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::appendBufferInternal):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::minimumBufferedTime const):
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::evictCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::evictCodedFrames):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::evictCodedFrames):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/256441@main">https://commits.webkit.org/256441@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46145ed7a23b56cb00a702a102c50c5f7c58bfe2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105348 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5104 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33780 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88153 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101179 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101431 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82382 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30810 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73643 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39515 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37205 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20385 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4452 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41267 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43310 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39637 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -401,7 +403,7 @@ void SourceBufferPrivate::removeCodedFrames(const MediaTime& start, const MediaT
     completionHandler();
 }
 
-void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximumBufferSize, const MediaTime& currentTime, const MediaTime& duration, bool isEnded)
```
If evictCodedFrames algorithm is not going to need the `duration` parameter, you may as well remove it.

calvaris replied:
I didn't check, I will, but I'd say it might be used in some other implementations, maybe in some related to GPU process.

ntrrgc replied:
It's true that this argument exists in the evictCodedFrames() method of SourceBufferPrivateRemote and RemoteSourceBufferProxy, but their only use in them is to pass it around through the IPC, which is even more reason to remove it if it's no longer necessary.

calvaris replied:
I changed this.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -412,98 +414,104 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
-        if (!isBufferFullFor(newDataSize, maximumBufferSize)) {
-            break;
+            MediaTime newRangeStart = buffered.minimumBufferedTime();
+            if (newRangeStart == rangeStart)
```
This would happen if and only if removedCodedFrames() fails to remove at least the first frame of the range. But why would that happen?

calvaris replied:
Depending on the iterations and how small the partition is, specially if you make the `removeCodedFrames` algorithm more conservative, this can happen. In that case, you want to bail out and continue to the next partition.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -412,98 +414,104 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
+    do {
+        MediaTime timeChunk = MediaTime(timeChunkAsMilliseconds, 1000);
+        MediaTime rangeStart = buffered.minimumBufferedTime();
+        MediaTime rangeEnd = rangeStart + timeChunk;
```
This declaration could be inside the while loop. I'll add to that, in case it's necessary, that in my opinion, a `while (true) {/* complex calculation */ if (...) break; }` is always preferrable to having to repeat code at the end of the while.

calvaris replied:
I prefer the other structure, but I can try this.

calvaris replied:
I don't have strong opinion, but IMHO, but readability is kinda subjective and IMHO, my original writing is more readable. I would suggest going back to it if you agree.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -412,98 +414,104 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
+        MediaTime timeChunk = MediaTime(timeChunkAsMilliseconds, 1000);
+        MediaTime rangeStart = buffered.minimumBufferedTime();
+        MediaTime rangeEnd = rangeStart + timeChunk;
+        MediaTime maximumRangeEnd = currentTime - timeChunk;
```
`maximumRangeEnd` and `timeChunk` could be const to show that they're not modified inside the while, but only redefined per iteration of the loop. That may feel superfluous when writing the algorithm but helps when reading it.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -412,98 +414,104 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
+            // end equal to the removal range start and end timestamp respectively.
+            MediaTime realRangeEnd = std::min(rangeEnd, maximumRangeEnd);
+            if (rangeStart >= realRangeEnd)
+                break;
```
This is the first and last read of `rangeEnd`. Therefore, we have no need for a `realRangeEnd` variable. We could instead inline the `min()` expression in the declaration of `rangeEnd`, also removing the need for the extra loop iteration guard here.

calvaris replied:
I'll check this

calvaris replied:
You need to inline this twice because you'd have to write it twice. IMHO, it is more readable like this.

Anyway, there are some more `const`s I can add.

ntrrgc replied:
Where would you have to write it twice?

> IMHO, it is more readable like this.

I agree doing things in two lines is more readable, but having a variable named *thing* and another called *realThing* living in the same scope is a footgun.

calvaris replied:
Maybe rename it to `clampedThing`

ntrrgc replied:
That is indeed better wording. Specific, conveys meaning. Still, having two variables that can be easily confused for each other, that often but not always hold the same value, for too long in scope, is still a footgun.

calvaris replied:
I prefer not to change it further. The name should be enough, IMHO.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -54,6 +54,8 @@ namespace WebCore {
 // new current time without triggering this early return.
 // FIXME(135867): Make this gap detection logic less arbitrary.
 static const MediaTime discontinuityTolerance = MediaTime(1, 1);
+static const unsigned evictionAlgorithmInitialTimeChunk = 30000;
+static const unsigned evictionAlgorithmTimeChunkLowThreshold = 3000;
```
Why milliseconds? Why not MediaTime? It's going to get converted to MediaTime anyway.

calvaris replied:
I guess at the moment of coding it, it was easier to divide by two. I could try to get it back to MediaTime.

calvaris replied:
Now that I checked, MediaTimes does not have support for any `operator/` so I could not use it unless I extracted the value from the MediaTime, divide it and recreate it. I find current solution better.

ntrrgc replied:
That is baffling to me as dividing MediaTimes should be trivial (just multiply the denominator). I'll have a look.

calvaris replied:
It is for fractions, maybe not so easy for doubles.

ntrrgc replied:
What would be the problem for doubles? Division in IEEE 754 handles division by zero even, so I can imagine it being a matter of dividing doubles and updating MediaTime flags.

Yesterday I started writing a patch for adding division to MediaTime (didn't get very far because it was late and soon got busy with other stuff), so I'm particularly interested in potential problems.

calvaris replied:
Sure, but I didn't think necessary to write it, just avoided the fuzz.

ntrrgc replied:
Oh yeah. You don't need to go into that rabbit hole for this patch. We have someone in-team for that. :rabbit2::raising_hand_woman:

ntrrgc reviewed (commented):
LGTM. I've proposed some cleanups.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -412,98 +414,98 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
+            size_t startTimeRange = buffered.find(rangeStart);
+            if (currentTimeRange != notFound && startTimeRange == currentTimeRange) {
+                size_t endTimeRange = buffered.find(rangeEnd);
+                if (currentTimeRange != notFound && endTimeRange == currentTimeRange)
```
`currentTimeRange != notFound` is a superfluous check, as that was checked in the containing if(). You may have meant `endTimeRange != notFound`, which incidentally is also superfluous.

ntrrgc reviewed (approved):
r+ with the mentioned nit

calvaris committed:
[MSE] Improved coded frame eviction algorithm
https://bugs.webkit.org/show_bug.cgi?id=246144

Reviewed by Alicia Boya Garcia.

I made it in two aspects, first is that now it does not begin at 0, it
begins at the first buffered range.

A fix would be that sometimes you need to evict frames but a window of
30s is too much and it can easily happen it does not evict anything,
so we need to reduce the partition to be able to effectively evict
something. We do it in windows from 30 to 3s.

This is tested by several tests that needed a rebase line or small tweaks.

* LayoutTests/media/media-source/media-source-append-before-last-range-no-quota-exceeded-expected.txt:
* LayoutTests/media/media-source/media-source-append-before-last-range-no-quota-exceeded.html:
* LayoutTests/media/media-source/media-source-append-buffer-full-quota-exceeded-error-expected.txt:
* LayoutTests/media/media-source/media-source-append-buffer-full-quota-exceeded-error.html:
* LayoutTests/media/media-source/media-source-evict-codedframe-after-seek-expected.txt:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::appendBufferInternal):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::minimumBufferedTime const):
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::evictCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::evictCodedFrames):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::evictCodedFrames):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: https://commits.webkit.org/256441@main
